### PR TITLE
Add userpace-probe support

### PR DESCRIPTION
**Pull request by frdeso**

### CODE REVIEW

File: lttng-events.c [hunk 1]
```diff
@@ -898,6 +999,15 @@ void _lttng_event_destroy(struct lttng_event *event)
 	case LTTNG_KERNEL_NOOP:
 	case LTTNG_KERNEL_SYSCALL:
 		break;
+	case LTTNG_KERNEL_UPROBE:
+		module_put(event->desc->owner);
+		lttng_uprobes_destroy_private(event);
+		break;
+	case LTTNG_KERNEL_URETPROBE:
+		/* 2 references were grabbed on the register()*/
```
Comment (compudj) [inline lttng-events.c]: missing space before */
missing . at end of sentence.

File: lttng-events.c [hunk 2]
```diff
@@ -754,6 +764,87 @@ struct lttng_event *_lttng_event_create(struct lttng_channel *chan,
...
+				event, event_return);
+		if (ret)
+			goto register_error;
+		/* Take 2 refs on the module: one per event. */
+		ret = try_module_get(event->desc->owner);
+		WARN_ON_ONCE(!ret);
+		ret = try_module_get(event->desc->owner);
+		WARN_ON_ONCE(!ret);
+		/*
+		 * Add the event_return to the metadata
```
Comment (compudj) [inline lttng-events.c]: missing . at end of sentence.

File: lttng-events.c [hunk 3]
```diff
@@ -754,6 +764,87 @@ struct lttng_event *_lttng_event_create(struct lttng_channel *chan,
...
+		ret = try_module_get(event->desc->owner);
+		WARN_ON_ONCE(!ret);
+		break;
+	case LTTNG_KERNEL_URETPROBE:
+	{
+		struct lttng_event *event_return;
+
+		/*
+		 * uretprobe defines 2 LTTng events: function entry and function
+		 * return
```
Comment (compudj) [inline lttng-events.c]: missing .

File: lttng-abi.h [hunk 4]
```diff
@@ -87,6 +89,14 @@ struct lttng_kernel_function_tracer {
 } __attribute__((packed));
 
 /*
+ * This structure is used for both uprobes and uretprobes
```
Comment (compudj) [inline lttng-abi.h]: missing .

File: probes/lttng-probe-utils.c [hunk 5]
```diff
@@ -62,3 +63,35 @@ long lttng_strlen_user_inatomic(const char *addr)
 	return count;
 }
 EXPORT_SYMBOL_GPL(lttng_strlen_user_inatomic);
+
+/*
+ * Returns the inode struct from the current task and an fd. The inode is grabed
+ * by this function and most be put once we are done with it using iput()
```
Comment (compudj) [inline probes/lttng-probe-utils.c]: missing .

Comment (compudj) [reply probes/lttng-probe-utils.c]: grabed -> grabbed

Comment (compudj) [reply probes/lttng-probe-utils.c]: most -> must

File: probes/lttng-probe-utils.c [hunk 6]
```diff
@@ -62,3 +63,35 @@ long lttng_strlen_user_inatomic(const char *addr)
...
+ */
+struct inode *lttng_get_inode_from_fd(int fd)
+{
+	struct file *file;
+	struct inode *inode;
+
+	rcu_read_lock();
+	/*
+	 * Returns the file backing the given fd. Needs to be done inside an RCU
+	 * critical section
```
Comment (compudj) [inline probes/lttng-probe-utils.c]: missing .

File: probes/lttng-probe-utils.c [hunk 7]
```diff
@@ -62,3 +63,35 @@ long lttng_strlen_user_inatomic(const char *addr)
...
+	 */
+	file = fcheck(fd);
+	if (file == NULL) {
+		printk(KERN_WARNING "Cannot access file backing the fd(%d)\n", fd);
+		inode = NULL;
+		goto error;
+	}
+
+	/* Grab a reference on the inode */
```
Comment (compudj) [inline probes/lttng-probe-utils.c]: missing .

File: probes/lttng-probe-utils.h [hunk 8]
```diff
@@ -21,10 +22,18 @@
...
+
 /*
  * Calculate string length. Include final null terminating character if there is
  * one, or ends at first fault.
  */
 long lttng_strlen_user_inatomic(const char *addr);
 
-#endif /* _LTTNG_PROBE_USER_H */
+/*
+ * Returns the inode struct from the current task and an fd. The inode is grabed
```
Comment (compudj) [inline probes/lttng-probe-utils.h]: grabed -> grabbed

File: probes/lttng-uprobes.c [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+	struct lttng_event *event =
+		container_of(uc, struct lttng_event, u.uprobe.up_consumer);
+	struct lttng_probe_ctx lttng_probe_ctx = {
+		.event = event,
+		.interruptible = !lttng_regs_irqs_disabled(regs),
+	};
+	struct lttng_channel *chan = event->chan;
+	struct lib_ring_buffer_ctx ctx;
+	int ret;
+
```
Comment (compudj) [inline probes/lttng-uprobes.c]: remove whiteline.

File: probes/lttng-uprobes.c [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+	payload.ip = regs->ip;
+
+	lib_ring_buffer_align_ctx(&ctx, lttng_alignof(payload));
+	chan->ops->event_write(&ctx, &payload, sizeof(payload));
+	chan->ops->event_commit(&ctx);
+	return 0;
+}
+
+/*
+ * Create event description
```
Comment (compudj) [inline probes/lttng-uprobes.c]: missing .

File: probes/lttng-uprobes.c [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+	desc->name = kstrdup(name, GFP_KERNEL);
+	if (!desc->name) {
+		ret = -ENOMEM;
+		goto error_str;
+	}
+
+	desc->nr_fields = 1;
+	desc->fields = fields =
+		kzalloc(1 * sizeof(struct lttng_event_field), GFP_KERNEL);
+
```
Comment (compudj) [inline probes/lttng-uprobes.c]: remove whiteline.

File: probes/lttng-uprobes.c [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+	ret = lttng_create_uprobe_event(name, event);
+	if (ret)
+		goto error;
+
+	inode = lttng_get_inode_from_fd(fd);
+	if (!inode) {
+		printk(KERN_WARNING "Cannot get inode from fd\n");
+		ret = -EBADF;
+		goto error;
```
Comment (compudj) [inline probes/lttng-uprobes.c]: I think we leak event->desc->name and event->desc on error here.

File: probes/lttng-uretprobes.c [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+	struct kref kref_inode;
+};
+
+static
+int _lttng_uretprobes_handler(struct uprobe_consumer *uc, struct pt_regs *regs,
+					 enum lttng_uretprobe_type type)
+{
+	struct lttng_urp *urp = container_of(uc, struct lttng_urp, up_consumer);
+	struct lttng_event *event = urp->event[type];
+
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: remove whiteline.

File: probes/lttng-uretprobes.c [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+	struct lttng_probe_ctx lttng_probe_ctx = {
+		.event = event,
+		.interruptible = !lttng_regs_irqs_disabled(regs),
+	};
+	struct lttng_channel *chan = event->chan;
+	struct lib_ring_buffer_ctx ctx;
+	int ret;
+
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: remove whiteline

File: probes/lttng-uretprobes.c [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+		return 0;
+
+	lib_ring_buffer_ctx_init(&ctx, chan->chan, &lttng_probe_ctx,
+				 sizeof(payload), lttng_alignof(payload), -1);
+
+	ret = chan->ops->event_reserve(&ctx, event->id);
+	if (ret < 0)
+		return 0;
+
+	/* event payload */
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: Event payload.

File: probes/lttng-uretprobes.c [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+	payload.ip = regs->ip;
+
+	lib_ring_buffer_align_ctx(&ctx, lttng_alignof(payload));
+	chan->ops->event_write(&ctx, &payload, sizeof(payload));
+	chan->ops->event_commit(&ctx);
+	return 0;
+}
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: add whiteline

File: probes/lttng-uretprobes.c [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+static
+int lttng_uretprobes_handler_return(struct uprobe_consumer *uc,
+					unsigned long func,
+					struct pt_regs *regs)
+{
+	return _lttng_uretprobes_handler(uc, regs, EVENT_RETURN);
+}
+
+/*
+ * Create event description
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: missing .

File: probes/lttng-uretprobes.c [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+ */
+static
+int lttng_create_uprobe_event(const char *name, struct lttng_event *event,
+					  enum lttng_uretprobe_type type)
+{
+	struct lttng_event_desc *desc;
+	struct lttng_event_field *fields;
+	int ret;
+	char *suffix;
+	char *alloc_name;
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: char *suffix, *alloc_name;

File: probes/lttng-uretprobes.c [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+	case EVENT_ENTRY:
+		suffix = "_entry";
+		break;
+	case EVENT_RETURN:
+		suffix = "_return";
+		break;
+	default:
+		ret = -1;
+		goto error;
+		break;
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: remove break (goto error is enough)

Comment (compudj) [reply probes/lttng-uretprobes.c]: we leak "desc" on error.

File: probes/lttng-uretprobes.c [hunk 20]
```diff
@@ -0,0 +1,325 @@
...
+			   struct lttng_event *event_entry,
+			   struct lttng_event *event_return)
+{
+	int ret;
+	struct inode *inode;
+	struct lttng_urp *lttng_urp;
+
+	/*
+	 * Create an event for both the entry and the return of the target
+	 * function
```
Comment (compudj) [inline probes/lttng-uretprobes.c]: missing .

File: wrapper/uprobes.h [hunk 21]
```diff
@@ -0,0 +1,88 @@
+#ifndef _LTTNG_WRAPPER_UPROBES_H
+#define _LTTNG_WRAPPER_UPROBES_H
+
+/*
+ * wrapper/uprobes.h
+ *
+ * wrapper around splice_to_pipe. Using KALLSYMS to get its address when
```
Comment (compudj) [inline wrapper/uprobes.h]: around uprobes....

File: wrapper/uprobes.h [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+#include <linux/version.h>
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,5,0))
+#include <linux/uprobes.h>
+
+/* Use kallsym lookup for version before 3.9 */
```
Comment (compudj) [inline wrapper/uprobes.h]: missing .

File: wrapper/uprobes.h [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
```
Comment (compudj) [inline wrapper/uprobes.h]: remove whiteline

File: wrapper/uprobes.h [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+        if (uprobe_unregister_sym) {
+                uprobe_unregister_sym(inode, offset, uc);
+        } else {
+                printk(KERN_WARNING "LTTng: uprobe_unregister symbol lookup failed.\n");
+                WARN_ON(1);
+        }
+}
+#endif
+#else
+/* Version <  3.5, before uprobe was added */
```
Comment (compudj) [inline wrapper/uprobes.h]: add .

File: lttng-events.c (hunk 3 above)
Comment (frdeso) [reply lttng-events.c]: Done.

File: lttng-events.c (hunk 2 above)
Comment (frdeso) [reply lttng-events.c]: Done.

File: lttng-abi.h (hunk 4 above)
Comment (frdeso) [reply lttng-abi.h]: Done.

File: probes/lttng-probe-utils.c (hunk 5 above)
Comment (frdeso) [reply probes/lttng-probe-utils.c]: Done.

File: probes/lttng-probe-utils.c (hunk 7 above)
Comment (frdeso) [reply probes/lttng-probe-utils.c]: Done.

File: probes/lttng-uprobes.c (hunk 9 above)
Comment (frdeso) [reply probes/lttng-uprobes.c]: Done.

File: probes/lttng-uprobes.c (hunk 10 above)
Comment (frdeso) [reply probes/lttng-uprobes.c]: Done.

File: probes/lttng-uprobes.c (hunk 11 above)
Comment (frdeso) [reply probes/lttng-uprobes.c]: Done.

File: probes/lttng-uretprobes.c (hunk 13 above)
Comment (frdeso) [reply probes/lttng-uretprobes.c]: Done.

File: probes/lttng-uretprobes.c (hunk 14 above)
Comment (frdeso) [reply probes/lttng-uretprobes.c]: Done.

File: probes/lttng-uretprobes.c (hunk 15 above)
Comment (frdeso) [reply probes/lttng-uretprobes.c]: Done.

File: probes/lttng-uretprobes.c (hunk 16 above)
Comment (frdeso) [reply probes/lttng-uretprobes.c]: Done.

File: probes/lttng-uretprobes.c (hunk 17 above)
Comment (frdeso) [reply probes/lttng-uretprobes.c]: Done.

File: probes/lttng-uretprobes.c (hunk 18 above)
Comment (frdeso) [reply probes/lttng-uretprobes.c]: Done.

File: wrapper/uprobes.h (hunk 21 above)
Comment (frdeso) [reply wrapper/uprobes.h]: Done.

File: wrapper/uprobes.h (hunk 23 above)
Comment (frdeso) [reply wrapper/uprobes.h]: Done.

File: wrapper/uprobes.h (hunk 22 above)
Comment (frdeso) [reply wrapper/uprobes.h]: done.

File: wrapper/uprobes.h (hunk 24 above)
Comment (frdeso) [reply wrapper/uprobes.h]: Done.

File: probes/lttng-uretprobes.c (hunk 19 above)
Comment (frdeso) [reply probes/lttng-uretprobes.c]: Done.

File: probes/lttng-uprobes.c (hunk 12 above)
Comment (frdeso) [reply probes/lttng-uprobes.c]: Done.

File: probes/lttng-probe-utils.h (hunk 8 above)
Comment (frdeso) [reply probes/lttng-probe-utils.h]: done.

**Comment by compudj** [comment]
merged manually, thanks !!

**Pull request closed without merging**